### ISSUE TITLE
os_server: stuck on auth parameters

### ISSUE THREAD

**Issue by Thingee** [ansible/ansible#19621]
_From @mohi048 on November 25, 2016 11:57_

##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

os_server

##### ANSIBLE VERSION

```
ansible 2.1.1.0
```

##### OS / ENVIRONMENT
ubuntu 14.04
Openstack Mitaka

##### SUMMARY

1. Tyring to create an instance via ansible
2. Getting "msg": "Error fetching server list on ********s:: (Inner Exception: Problem with auth parameters)"}

##### STEPS TO REPRODUCE
1. Run the playbook 

```
ansible-playbook test-play.yml -i hosts -vvv 


---
- name: Provision a compute instance
  gather_facts: no
  become: no
  hosts: 06-deploy-openstack
  user: root

  tasks:

    - name: Set necessary variables when Tower cloud credential config is present
      set_fact:
        os_auth:
          auth_type: password
          auth_url: http://192.168.56.104:35357/v3
          username: admin
          password: Secrete
          project_name: admin
          os_tenant_name: admin
          project_domain_name: default
          availability_zone: nova
          user_domain_name: default
          region_name: RegionOne
          auth_version: 2
          compute_api_version: 2


    - name: Launch compute instances
      os_server:
        auth: "{{ os_auth }}"
        name: compute-node-demo
        auto_ip: no
        image: Cirros
        flavor: m1.tiny
        availability_zone: nova
        nics:
          - net-name: public
        wait: no
        state: present



```

##### EXPECTED RESULTS

Should create the instance
##### ACTUAL RESULTS

```
No config file found; using defaults

PLAYBOOK: test-play.yml ********************************************************
1 plays in test-play.yml

PLAY [Provision a compute instance] ********************************************

TASK [Set necessary variables when Tower cloud credential config is present] ***
task path: /Users/cccc/my_proj/Ansible/OpenStack/All-In-One_Liberty/test-play.yml:21
ok: [192.168.56.104] => {"ansible_facts": {"os_auth": {"auth_type": "password", "auth_url": "http://192.168.56.104:35357/v3", "auth_version": 2, "availability_zone": "nova", "compute_api_version": 2, "os_tenant_name": "admin", "password": "Secrete", "project_domain_name": "default", "project_name": "admin", "user_domain_name": "default", "username": "admin"}}, "changed": false, "invocation": {"module_args": {"os_auth": {"auth_type": "password", "auth_url": "http://192.168.56.104:35357/v3", "auth_version": 2, "availability_zone": "nova", "compute_api_version": 2, "os_tenant_name": "admin", "password": "Secrete", "project_domain_name": "default", "project_name": "admin", "user_domain_name": "default", "username": "admin"}}, "module_name": "set_fact"}}

TASK [Launch compute instances] ************************************************
task path: /Users/cccc/my_proj/Ansible/OpenStack/All-In-One_Liberty/test-play.yml:55
<192.168.56.104> ESTABLISH CONNECTION FOR USER: root on PORT 22 TO 192.168.56.104
<192.168.56.104> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1480074461.66-185801142654384 `" && echo ansible-tmp-1480074461.66-185801142654384="` echo $HOME/.ansible/tmp/ansible-tmp-1480074461.66-185801142654384 `" ) && sleep 0'
<192.168.56.104> PUT /var/folders/pf/r6d2yn7d2h1c1m6ttmdvct9msv0smb/T/tmpCqkVXK TO /root/.ansible/tmp/ansible-tmp-1480074461.66-185801142654384/os_server
<192.168.56.104> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1480074461.66-185801142654384/os_server; rm -rf "/root/.ansible/tmp/ansible-tmp-1480074461.66-185801142654384/" > /dev/null 2>&1 && sleep 0'
fatal: [192.168.56.104]: FAILED! => {"changed": false, "extra_data": null, "failed": true, "invocation": {"module_args": {"api_timeout": null, "auth": {"auth_type": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "auth_url": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "auth_version": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "availability_zone": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "compute_api_version": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "os_tenant_name": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "project_domain_name": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "project_name": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "user_domain_name": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"}, "auth_type": null, "auto_ip": false, "availability_zone": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "boot_from_volume": false, "boot_volume": null, "cacert": null, "cert": null, "cloud": null, "config_drive": false, "endpoint_type": "public", "flavor": "m1.tiny", "flavor_include": null, "flavor_ram": null, "floating_ip_pools": null, "floating_ips": null, "image": "Cirros", "image_exclude": "(deprecated)", "key": null, "key_name": null, "meta": null, "name": "compute-node-demo", "network": null, "nics": [{"net-name": "public"}], "region_name": null, "scheduler_hints": null, "security_groups": ["VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"], "state": "present", "terminate_volume": false, "timeout": 180, "userdata": null, "verify": true, "volume_size": false, "volumes": [], "wait": false}, "module_name": "os_server"}, "msg": "Error fetching server list on ********s:: (Inner Exception: Problem with auth parameters)"}

NO MORE HOSTS LEFT *************************************************************
 [WARNING]: Could not create retry file 'test-play.retry'.         [Errno 2] No such file or directory: ''


PLAY RECAP *********************************************************************
192.168.56.104             : ok=1    changed=0    unreachable=0    failed=1   


```

_Copied from original issue: ansible/ansible-modules-core#5734_

**Comment by Thingee** [comment]
_From @mohi048 on November 25, 2016 12:19_

Some additional info -:
Sourced openstack credentials file


[root@linux]# python -m os_client_config.config
('envvars', '', {'auth_type': u'password', u'compute_api_version': u'2', 'key': None, u'database_api_version': u'1.0', u'metering_api_version': u'2', 'auth_version': '3', u'network_api_version': u'2', u'message': u'', u'image_format': u'qcow2', 'networks': [], u'image_api_version': u'2', 'verify': True, u'dns_api_version': u'2', u'object_store_api_version': u'1', u'status': u'active', 'region_name': '', 'api_timeout': None, u'baremetal_api_version': u'1', 'auth': {'username': 'admin', 'project_name': 'admin', 'user_domain_name': 'default', 'auth_url': 'http://192.168.56.104:35357/v3', 'password': 'Secrete', 'project_domain_name': 'default'}, u'image_api_use_tasks': False, u'floating_ip_source': u'neutron', u'orchestration_api_version': u'1', 'cacert': None, u'key_manager_api_version': u'v1', u'identity_api_version': '3', u'volume_api_version': '2', 'cert': None, u'secgroup_source': u'neutron', u'container_api_version': u'1', u'interface': u'public', u'disable_vendor_agent': {}})


>>> import shade
>>> shade.simple_logging(debug=True)
>>> cloud = shade.openstack_cloud()
>>> my_servers = cloud._list_servers()
>>> my_servers

[Munch({'vm_state': u'stopped', 'OS-EXT-STS:task_state': None, 'addresses': {u'public': [{u'OS-EXT-IPS-MAC:mac_addr': u'fa:16:3e:ba:cd:82', u'version': 4, u'addr': u'192.168.56.22', u'OS-EXT-IPS:type': u'fixed'}]}, 'terminated_at': None, 'image': {u'id': u'38f423b7-3964-40bb-9365-5bdf90dd9e7f'}, 'OS-EXT-AZ:availability_zone': u'nova', 'OS-EXT-STS:vm_state': u'stopped', 'OS-EXT-SRV-ATTR:instance_name': u'instance-00000010', 'OS-SRV-USG:launched_at': u'2016-11-24T16:54:12.000000', 'flavor': {u'id': u'1'}, 'az': u'nova', 'id': u'80183245-0ac7-4cbd-90ab-b751db1a8a0d', 'security_groups': [{u'name': u'default'}], 'os-extended-volumes:volumes_attached': [], 'user_id': u'f91e1b20a08447a196c4fba8eb74e473', 'disk_config': u'AUTO', 'OS-DCF:diskConfig': u'AUTO', 'networks': {u'public': [u'192.168.56.22']}, 'accessIPv4': u'192.168.56.22', 'accessIPv6': '', 'location': Munch({'project': Munch({'domain_id': None, 'id': u'5d893ad252ca4f0885ef250f21113953', 'name': 'admin', 'domain_name': 'default'}), 'zone': u'nova', 'region_name': '', 'cloud': 'envvars'}), 'power_state': 4, 'public_v4': u'192.168.56.22', 'progress': 0, 'OS-EXT-STS:power_state': 4, 'interface_ip': u'192.168.56.22', 'launched_at': u'2016-11-24T16:54:12.000000', 'metadata': {}, 'status': u'SHUTOFF', 'updated': u'2016-11-25T15:57:30Z', 'hostId': u'5f764837974fe1e734c5e0acf4c98ecf3f66b6dcb7c58c7684b67201', 'OS-EXT-SRV-ATTR:host': u'Mitaka', 'OS-SRV-USG:terminated_at': None, 'key_name': None, 'public_v6': '', 'request_ids': [], 'private_v4': '', 'cloud': 'envvars', 'host_id': u'5f764837974fe1e734c5e0acf4c98ecf3f66b6dcb7c58c7684b67201', 'task_state': None, 'properties': {'OS-EXT-STS:task_state': None, 'OS-EXT-SRV-ATTR:host': u'Mitaka', 'OS-SRV-USG:terminated_at': None, 'OS-DCF:diskConfig': u'AUTO', 'request_ids': [], 'os-extended-volumes:volumes_attached': [], 'OS-EXT-STS:vm_state': u'stopped', 'OS-EXT-SRV-ATTR:instance_name': u'instance-00000010', 'OS-SRV-USG:launched_at': u'2016-11-24T16:54:12.000000', 'OS-EXT-SRV-ATTR:hypervisor_hostname': u'Mitaka', 'OS-EXT-STS:power_state': 4, 'OS-EXT-AZ:availability_zone': u'nova'}, 'project_id': u'5d893ad252ca4f0885ef250f21113953', 'OS-EXT-SRV-ATTR:hypervisor_hostname': u'Mitaka', 'name': u'cvb', 'adminPass': None, 'tenant_id': u'5d893ad252ca4f0885ef250f21113953', 'region': '', 'created': u'2016-11-24T16:53:54Z', 'has_config_drive': False, 'volumes': [], 'config_drive': u''})]

**Comment by Thingee** [comment]
_From @ansibot on November 25, 2016 12:23_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, @thingee, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
_From @ansibot on December 7, 2016 20:41_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by gyurco** [comment]
Exception: Problem with auth parameters
is because of auth_version: 2 and compute_api_version: 2 are _not_ valid auth plugin parameters. They must be given before the plugin is instantiated. However the problem is that there is no (documented) way to do it, so I'm also stuck with the os_xxx ansible modules. The default Keystone API version is 2.0, and cannot find a way how the specify 3 in the playbook (not in cloud.yaml).

**Comment by jvanstone** [comment]
Is there any news on this? I'm also running into this problem. It seems to be something recent? I had a playbook that was working in mid December that now fails with this error.

**Comment by emonty** [comment]
Hey all - I've made this: https://github.com/ansible/ansible/pull/20974 which should provide a mechanism to provide all of the parameters one can provide in clouds.yaml directly in a playbook. It means you'd want to change the initial example to:

```yaml
---
- name: Provision a compute instance
  gather_facts: no
  become: no
  hosts: 06-deploy-openstack
  user: root

  tasks:

    - name: Set necessary variables when Tower cloud credential config is present
      set_fact:
        os_cloud:
          auth:
              auth_url: http://192.168.56.104:35357/v3
              username: admin
              password: Secrete
              project_name: admin
              project_domain_name: default
              user_domain_name: default
          auth_type: password
          region_name: RegionOne
          auth_version: 2
          compute_api_version: 2
    - name: Launch compute instances
      os_server:
        cloud: "{{ os_cloud }}"
        name: compute-node-demo
        auto_ip: no
        image: Cirros
        flavor: m1.tiny
        availability_zone: nova
        nics:
          - net-name: public
        wait: no
        state: present
```

**Comment by ssbarnea** [comment]
I can confirm this bug even with latest code from `devel` branch and it is very easy to reproduce.

This works and proves that shade is correctly configured:

```
import shade
shade.simple_logging(debug=True)
cloud = shade.openstack_cloud()
my_servers = cloud._list_servers()
print(my_servers)
```
As a detail `~/.config/openstack/clouds.yml` contains multiple clouds and we do also have an environment variable which defines the default cloud to use `OS_CLOUD=myclould`.

Now back to ansible shade(d) modules, here is what happens with minimal test case:
```
#!/usr/bin/env ansible-playbook -i localhost,
---
- name: test output
  hosts: localhost
  gather_facts: no
  vars:
      ansible_connection: local
  tasks:
    - name: os_network module testing
      os_network:
        name: some_non_existing_network
        state: absent
```

Execution result:
```
TASK [os_network module testing] **********************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "failed": true, "msg": "Error fetching network list: Problem with auth parameters"}
```
There are multiple problem involved here:
- it should not fail
- even if it fails the error message should contain useful information, `Problem with auth parameters` is bit far from *clear*.

Note: It seems that if I specify `cloud: mycloud` the command will succeed, so what happens here is that the environment variable OS_CLOUD does not reach shade code and shade fails to determine which cloud to use.

**Comment by ssbarnea** [comment]
Update: It seems that the reasons why OS_CLOUD didn't reach shade was because I was running via tox which was not configured to pass environment variables.

I can confirm that if defined OS_CLOUD is used.

Still, I do find the error confusing and I would have prefered to just try to use the first found cloud in the config or to throw a clear error like "unable to determine which cloud to use out of x defined".

**Comment by ansibot** [comment]
cc @dagnello
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @omgjlk
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @mnaser
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mnaser** [comment]
wontfix

The default behaviour that you must either provide credentials or provide a cloud.  I disagree that we should have to default to the first one because that might have some bad implications in terms of changing behaviour.
